drivers/sds011/sds011.c

```diff
@@ -0,0 +1,351 @@
+
+    for (unsigned i = 0; i < len; i++) {
+        cmd[SDS011_DB1_IDX + i] = data_bytes[i];
+        checksum += data_bytes[i];
+    }
+
+    cmd[SDS011_DEVID1_IDX] = (dev->params.dev_id >> 8) & 0xFF;
+    checksum += cmd[SDS011_DEVID1_IDX];
+    cmd[SDS011_DEVID2_IDX] = dev->params.dev_id & 0xFF;
+    checksum += cmd[SDS011_DEVID2_IDX];
+
+    cmd[SDS011_FRAME_SEND_LEN - 2] = checksum & SDS011_FRAME_CSUM_MSK;
+    cmd[SDS011_FRAME_SEND_TAIL_IDX] = SDS011_FRAME_TAIL;
+
+    mutex_lock(&dev->dev_lock);
+
+    dev->pos = 0;
+    dev->checksum = 0;
+
+    mutex_lock(&dev->cb_lock);
+
+    /* if no active reporting callback is registered, UART must be enabled first */
+    if((dev->cb == NULL) &&
+       (uart_init(dev->params.uart, SDS011_UART_BAUDRATE, _rx_cb, dev) != 0)) {
+        mutex_unlock(&dev->cb_lock);
+        mutex_unlock(&dev->dev_lock);
+        return SDS011_ERROR;
+    }
+
+    uart_write(dev->params.uart, cmd, SDS011_FRAME_SEND_LEN);
+
+    /* wait for the isr callback to unlock the mutex */
+    mutex_lock(&dev->cb_lock);
+
+    /* only copy data when checksum was valid */
+    if (dev->pos != 0) {
+        memcpy(recv_frm, dev->rx_mem, SDS011_FRAME_RECV_LEN);
+        /* mark the recv buffer as free */
+        dev->pos = 0;
+        dev->checksum = 0;
+
+        /* check if we received a valid response for the cmd sent*/
+        if(((recv_frm[SDS011_CMDID_IDX]  == SDS011_RCMDID_REPLY) &&
+            (cmd[SDS011_DB1_IDX] == recv_frm[SDS011_DB1_IDX]))
+           || ((recv_frm[SDS011_CMDID_IDX] == SDS011_RCMDID_DATA)
+            && (cmd[SDS011_DB1_IDX] == SDS011_CMD_DB1_QUERY_DATA))) {
+            res = SDS011_OK;
+        }
+        else {
+            res = SDS011_INVALID_RESPONSE;
+        }
+    }
+    else {
+        res = SDS011_INVALID_CHKSUM;
+    }
+
+    /* reset mutex state */
+    mutex_unlock(&dev->cb_lock);
+
+    /* if no active reporting callback is registered, UART can be disabled */
+    if((dev->cb == NULL) &&
+       (uart_init(dev->params.uart, SDS011_UART_BAUDRATE, NULL, NULL) != 0)) {
+        res = SDS011_ERROR;
+    }
+
+    /* release device */
+    mutex_unlock(&dev->dev_lock);
+
+    return res;
+}
+
+/** @brief shorthand for getting a single byte property with _send_recv_cmd  */
+static int _get_property(sds011_t *dev, uint8_t *cmd, size_t len,
+                         uint8_t *p, uint8_t p_idx)
+{
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, len, recv);
+
+    if (res == SDS011_OK) {
+        *p = recv[p_idx];
+    }
+
+    return res;
+}
+
+int sds011_init(sds011_t *dev, const sds011_params_t *params)
+{
+    assert(params->uart < UART_NUMOF);
+
+    if ((params->pwr_pin != GPIO_UNDEF) &&
+        (gpio_init(params->pwr_pin, GPIO_OUT) != 0)) {
+        return SDS011_ERROR;
+    }
+
+    memcpy(&dev->params, params, sizeof(sds011_params_t));
+
+    mutex_init(&dev->dev_lock);
+    mutex_init(&dev->cb_lock);
+
+    dev->cb = NULL;
+
+    sds011_power_on(dev);
```

Comment:
why? does this make sense without a callback?

Reply:
Yes because the device needs to be started up before you can request data and the driver guidelines say a device should be ready to interact with after the init call. Also, the driver (device) provides two different modes of operation. One where you need to request a measurement and one where the sensor reports periodically. The driver defaults to requesting mode so the user is not forced to register a callback to talk to the device. You can still chose to use a callback if preferred - the driver can handle both.